apps/worker/tasks/base.py

```diff
@@ -212,13 +212,33 @@ def apply_async(self, args=None, kwargs=None, **options):
 
         # Pass current time in task headers so we can emit a metric of
         # how long the task was in the queue for
+        # Also track attempts to catch visibility timeout re-deliveries
         current_time = datetime.now()
         headers = {
             **opt_headers,
             "created_timestamp": current_time.isoformat(),
+            "attempts": 1,  # Initialize attempts counter
         }
         return super().apply_async(args=args, kwargs=kwargs, headers=headers, **options)
 
+    def retry(self, max_retries=None, countdown=None, exc=None, **kwargs):
+        """Override Celery's retry to always update attempts header.
+
+        This ensures all retries (direct calls or through safe_retry) track attempts
+        consistently, including re-deliveries from visibility timeout expiration.
+        """
+        headers = {}
+        request_headers = self._get_request_headers()
+        if request_headers:
+            headers.update(request_headers)
+        headers.update(kwargs.get("headers", {}) or {})
+        headers["attempts"] = self.attempts + 1
+        kwargs["headers"] = headers
+
+        return super().retry(
+            max_retries=max_retries, countdown=countdown, exc=exc, **kwargs
+        )
```

Comment:
this function and the one below is doing a lot of protection code, can this not be simplified? the only real thing that we're doing is pulling the attempts and incrementing it